File: apps/sonar_led/sonar_led.py
#!/usr/bin/env python
import socket
import sys
import time

# This program reads values from the sonar and displays
# them on the leds.

def set_cmd(sock, set_str):
    sock.send(set_str)
    while True:
        retval = sock.recv(1)
        if retval == '\\':
            break

def get_cmd(sock, get_str):
    sock.send(get_str)
    data = ""
    while True:
        retval = sock.recv(1)
        if retval == '\\':
            break
        elif retval == '\n':
            pass
        else:
            data = data + retval

    return data


RANGE = 3
count = 0;

try:
    sock_sonar = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock_sonar.connect(('localhost', 8870))

    sock_led = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock_led.connect(('localhost', 8870))

    # Make sure we are talking to the correct serial port
    # set_cmd(sock_sonar,'hbaset serial_fpga port /dev/ttyUSB1\n')

    # Enable both sonars
    set_cmd(sock_sonar,'hbaset hba_sonar ctrl 3\n')

    time.sleep(0.5)

    # loop forever 
    while True:
        dist_l_str = get_cmd(sock_sonar,'hbaget hba_sonar sonar0\n')
        dist_l = int(dist_l_str, 16)

        dist_r_str = get_cmd(sock_sonar,'hbaget hba_sonar sonar1\n')
        dist_r = int(dist_r_str, 16)


        print "dist_l: %d,  dist_r: %d" % (dist_l, dist_r)

        if dist_l < (RANGE*1):
            set_cmd(sock_led,'hbaset hba_basicio leds ' "%x" '\n' % 0x01)
        elif dist_l < (RANGE*2):
            set_cmd(sock_led,'hbaset hba_basicio leds ' "%x" '\n' % 0x03)
        elif dist_l < (RANGE*3):
            set_cmd(sock_led,'hbaset hba_basicio leds ' "%x" '\n' % 0x07)
        elif dist_l < (RANGE*4):
            set_cmd(sock_led,'hbaset hba_basicio leds ' "%x" '\n' % 0x0f)
        elif dist_l < (RANGE*5):
            set_cmd(sock_led,'hbaset hba_basicio leds ' "%x" '\n' % 0x1f)
        elif dist_l < (RANGE*6):
            set_cmd(sock_led,'hbaset hba_basicio leds ' "%x" '\n' % 0x3f)
        elif dist_l < (RANGE*7):
            set_cmd(sock_led,'hbaset hba_basicio leds ' "%x" '\n' % 0x7f)
        else:
            set_cmd(sock_led,'hbaset hba_basicio leds ' "%x" '\n' % 0xff)

        time.sleep(0.1)


except KeyboardInterrupt:
    # exit on Ctrl^C
    sock_sonar.close()
    sock_led.close()
    sys.exit()

except socket.error:
    print "Couldn't connect to hbaserver"
    sys.exit()

File: apps/sonar_led/sonar_led.c
/* sonar_led.c  :  This program demonstrates the use of the
 * sonar and leds on the hba class project.
 *
 * Build with: gcc -o sonar_led sonar_led.c
 * Be sure hbaserver is running and listening on port 8870
 */


#include <stdio.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <stddef.h>
#include <string.h>    /* for memset */
#include <arpa/inet.h> /* for inet_addr() */


static void sndcmd(int fd, char *cmd); // send a command to the board, get prompt

int main()
{
    // XXX int8_t counter;         // the 8-bit count to display
    int  tmp_int;           // a temporary integer
    int  cmdfd;             // FD for commands for leds
    int  evtfd;             // FD for sonar data
    struct sockaddr_in skt; // network address for hbaserver
    int  adrlen;
    char strled[99];        // command to set the leds
    char strevt[99];        // where to receive the button press string
    int  sonar_val;           // latest button event as an integer
    int  RANGE=3;

    // Open connection to DPserver daemon
    adrlen = sizeof(struct sockaddr_in);
    (void) memset((void *) &skt, 0, (size_t) adrlen);
    skt.sin_family = AF_INET;
    skt.sin_port = htons(8870);
    if ((inet_aton("127.0.0.1", &(skt.sin_addr)) == 0) ||
        ((cmdfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) ||
        (connect(cmdfd, (struct sockaddr *) &skt, adrlen) < 0)) {
        printf("Error: unable to connect to hbaserver.\n");
        exit(-1);
    }

    /* Setup peripherals */
    // XXX sndcmd(cmdfd, "hbaset serial_fpga port /dev/ttyUSB1\n");
    sndcmd(cmdfd, "hbaset hba_sonar ctrl 1\n");
    sleep(1);

    /* Blink the LEDs */
    sndcmd(cmdfd, "hbaset hba_basicio leds ff\n");
    sleep(0.2);
    sndcmd(cmdfd, "hbaset hba_basicio leds 00\n");
    sleep(0.2);
    sndcmd(cmdfd, "hbaset hba_basicio leds ff\n");
    sleep(0.2);
    sndcmd(cmdfd, "hbaset hba_basicio leds 00\n");

    /* Open another connection to sonar data */
    (void) memset((void *) &skt, 0, (size_t) adrlen);
    skt.sin_family = AF_INET;
    skt.sin_port = htons(8870);
    if ((inet_aton("127.0.0.1", &(skt.sin_addr)) == 0) ||
        ((evtfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) ||
        (connect(evtfd, (struct sockaddr *) &skt, adrlen) < 0)) {
        printf("Error: unable to connect to hbaserver.\n");
        exit(-1);
    }

    // XXX counter = 0;   // leds are already showing zero

    /* Start the stream of button events */
    // write(evtfd, "dpcat hba_basicio buttons\n", 20);
    /* the above command never returns so we do not use sndcmd() */

    while(1) {
        /* read sonar data */
        sprintf(strevt, "hbaget hba_sonar sonar0\n");
        write(evtfd, strevt, strlen(strevt));
        read(evtfd, strevt, 4);    // two digits, newline, prompt
        sscanf(strevt, "%02x\n\\", &sonar_val);
        printf("sonar_val: %02x\n", sonar_val);

        /* display new value of count */
        // XXX sprintf(strled, "hbaset hba_basicio leds %02x\n", (counter & 0xff));
        if (sonar_val < RANGE*1) {
            sprintf(strled, "hbaset hba_basicio leds 1\n");
        } else if (sonar_val < RANGE*2) {
            sprintf(strled, "hbaset hba_basicio leds 3\n");
        } else if (sonar_val < RANGE*3) {
            sprintf(strled, "hbaset hba_basicio leds 7\n");
        } else if (sonar_val < RANGE*4) {
            sprintf(strled, "hbaset hba_basicio leds f\n");
        } else if (sonar_val < RANGE*5) {
            sprintf(strled, "hbaset hba_basicio leds 1f\n");
        } else if (sonar_val < RANGE*6) {
            sprintf(strled, "hbaset hba_basicio leds 3f\n");
        } else if (sonar_val < RANGE*7) {
            sprintf(strled, "hbaset hba_basicio leds 7f\n");
        } else {
            sprintf(strled, "hbaset hba_basicio leds ff\n");
        }

        sndcmd(cmdfd, strled);
        sleep(0.1);
        // XXX counter++;
    }
}

/* sndcmd():  Send a command to hbaserver and wait for a response.  The
 *     response will be a prompt character, which we ignore and return,
 *     or an error message which we send to stderr. */
static void sndcmd(int fd, char *cmd)
{
    size_t count;          // number of chars in command to send
    char   c;              // prompt or error message character
    int    retval;         // return value of read()

    count = strlen(cmd);        // should sanity check count
    write(fd, cmd, count);   // should look at write() return value

    /* loop getting characters.  Return on a prompt character '\' and
     * send any other character to stderr. */
    while (1) {
        retval = read(fd, &c, 1);
        if (0 >= retval)
            exit(1);       // did TCP conn go down?
        else if ('\\' == c)
            return;        // got a prompt char.  Done with command
        else
            write(2, &c, 1);    // send to stderr
    }
}
